staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -0,0 +1,192 @@
+}
+
+func validator(s *schema.Structural, isResourceRoot bool) *Validator {
+	compiledRules, err := Compile(s, isResourceRoot)
+	if s.Items != nil {
+		if i := validator(s.Items, s.Items.XEmbeddedResource); i != nil {
+			return &Validator{compiledRules: compiledRules, compilationErr: err, isResourceRoot: isResourceRoot, Items: i}
+		}
+	}
+	if len(s.Properties) > 0 {
+		properties := make(map[string]Validator, len(s.Properties))
+		for k, prop := range s.Properties {
+			if p := validator(&prop, prop.XEmbeddedResource); p != nil {
+				properties[k] = *p
+			}
+		}
+		if len(properties) > 0 {
+			return &Validator{compiledRules: compiledRules, compilationErr: err, isResourceRoot: isResourceRoot, Properties: properties}
+		}
+	}
+	if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+		if ap := validator(s.AdditionalProperties.Structural, s.AdditionalProperties.Structural.XEmbeddedResource); ap != nil {
+			return &Validator{compiledRules: compiledRules, compilationErr: err, isResourceRoot: isResourceRoot, AdditionalProperties: ap}
+		}
+	}
+	if len(compiledRules) > 0 {
+		return &Validator{compiledRules: compiledRules, compilationErr: err, isResourceRoot: isResourceRoot}
+	}
+
+	return nil
+}
+
+// Validate validates all x-kubernetes-validations rules in Validator against obj and returns any errors.
+func (s *Validator) Validate(fldPath *field.Path, sts *schema.Structural, obj interface{}) field.ErrorList {
+	if s == nil || obj == nil {
+		return nil
+	}
+
+	errs := s.validateExpressions(fldPath, sts, obj)
+	switch obj := obj.(type) {
+	case []interface{}:
+		return append(errs, s.validateArray(fldPath, sts, obj)...)
+	case map[string]interface{}:
+		return append(errs, s.validateMap(fldPath, sts, obj)...)
+	}
+	return errs
+}
+
+func (s *Validator) validateExpressions(fldPath *field.Path, sts *schema.Structural, obj interface{}) (errs field.ErrorList) {
+	if s.isResourceRoot {
+		sts = model.WithTypeAndObjectMeta(sts)
+	}
+	activation := NewValidationActivation(obj, sts)
+	if s.compilationErr != nil {
+		errs = append(errs, field.Invalid(fldPath, obj, fmt.Sprintf("rule compiler initialization error: %v", s.compilationErr)))
+		return errs
+	}
+	for i, compiled := range s.compiledRules {
+		rule := sts.XValidations[i]
+		if compiled.Error != nil {
+			errs = append(errs, field.Invalid(fldPath, obj, fmt.Sprintf("rule compile error: %v", compiled.Error)))
+			continue
+		}
+		evalResult, _, err := compiled.Program.Eval(activation)
+		if err != nil {
+			errs = append(errs, field.Invalid(fldPath, obj, fmt.Sprintf("runtime error for rule %s: %v", toJsonString(rule.Rule), err)))
+			continue
+		}
+		if evalResult != types.True {
+			if len(rule.Message) != 0 {
+				errs = append(errs, field.Invalid(fldPath, obj, rule.Message))
+			} else {
+				errs = append(errs, field.Invalid(fldPath, obj, fmt.Sprintf("failed rule: %s", toJsonString(rule.Rule))))
+			}
+		}
+	}
+	return errs
+}
+
+func toJsonString(rule string) string {
```

Comment:
I think a json-encoded rule is going to be _really_ hard to read... can we require a message if the rule is multi-line (maybe after trimming leading/trailing spaces), and output single-line rules directly in error messages?

Reply:
I've gone with: "Message represents the message displayed when validation failed. The message is required if the Rule contains line breaks. The message must not contain line breaks." and trimmed strings every it seemed appropriate.